torch/csrc/jit/passes/autocast.cpp

```diff
@@ -81,21 +81,49 @@ c10::optional<AutocastScope> parseAutocast(Value* value) {
 void castTensorInputs(Node* node, Symbol cast_op) {
   const auto graph = node->owningGraph();
 
-  WithInsertPoint insert_point(node);
-
   std::unordered_set<Value*> casted_inputs;
   for (auto input : node->inputs()) {
     if (input->type()->kind() == TensorType::Kind) {
       casted_inputs.insert(input);
     }
   }
 
+  WithInsertPoint insert_point(node);
+
   for (auto input : casted_inputs) {
     const auto new_input = graph->insert(cast_op, {input});
     node->replaceInputWith(input, new_input);
   }
 }
 
+void updateDefaultDtype(Node* node, at::ScalarType dtype) {
+  const auto graph = node->owningGraph();
+
+  const auto& actual_args = node->inputs();
+  const auto& formal_args = node->schema().arguments();
+  TORCH_INTERNAL_ASSERT(actual_args.size() == formal_args.size());
+
```

Comment:
The eager [fp32_set_opt_dtype wrapper](https://github.com/pytorch/pytorch/blob/b8e6e2971cba779daf14652d05f427544ab635b5/aten/src/ATen/autocast_mode.cpp#L145-L151) makes a semi-hacky attempt to bail out of setting the opt dtype if the first argument is not eligible for casting (ie if it's fp64, or not cuda). Do we also want to do that here?

Reply:
Yes, this is along the lines of the @jjsjann123's comments. Adding the runtime check the same way eager mode does is not trivial (since we're not wrapping the functions), but I think there's a way to get the same results by reusing the fp32 policy.

I just updated the logic, please take another look.

Reply:
hmm...`autocast_to_fp32` [here](aten::autocast_to_fp32) only acts on fp16 tensors (ignores fp64 tensors)...and we only trigger `autocast_to_fp32` if there's not an explicit dtype set...that seems good for ignoring fp64 inputs.

The thing that still confuses me is, calling `autocast_to_fp32` on the inputs implements the `fp32` policy, not the `fp32_set_opt_dtype` policy (which doesn't cast any inputs, but instead sets the dtype arg to fp32). So the control flow here is effectively a new policy: "`fp32`-like behavior if output dtype is unset."

Honestly, if we can't check the types of any Tensor args, i'm no longer sure it's possible to imitate eager mode's `fp32_set_opt_dtype` (including its bailout).

Reply:
Your summary is accurate: we're effectively reusing the fp32 policy + excluding the ops with explicit dtype argumens as an optimization. 

Personally, I think the unification of policies is a good thing (as long as it's feasible)

Reply:
My point is
> excluding the ops with explicit dtype argumens as an optimization.

is not the same optimization that the eager `fp32_set_opt_dtype` is doing. The eager version sets the dtype and defers to the op itself to cast the output to the right type. I don't think it's a big deal as long as we understand what's happening.

Reply:
Yes - it's not the exact same path, but same end result, right? At least the closest approximation I can think of with scripting.